libtenzir/builtins/operators/sample.cpp

```diff
@@ -0,0 +1,215 @@
+          auto b = int64_type::make_arrow_builder(arrow::default_memory_pool());
+          check(b->Reserve((rows + 1) / stride));
+          for (auto i = offset % stride; i < rows; i += stride) {
+            check(b->Append(i));
+          }
+          return finish(*b);
+        };
+    for (const auto& slice : input) {
+      if (auto now = std::chrono::steady_clock::now();
+          now - last > args_.period.inner) {
+        if (count > 1) {
+          // `count` is a `uint64_t` > 1 so `logx(count)` or `sqrt(count)` can
+          // never give a negative result.
+          stride = count > args_.min_events.value_or(0) ? std::ceil(calc()) : 1;
+        }
+        last = now - (now - last) % args_.period.inner;
+        offset = 0;
+        count = 0;
+      }
+      if (slice.rows() == 0
+          or (args_.max_rate and args_.max_rate.value() <= count)) {
+        co_yield {};
+        continue;
+      }
+      count += slice.rows();
+      auto batch = to_record_batch(slice);
+      const auto rows = batch->num_rows();
+      auto stride_index = make_stride_index(offset, rows, stride);
+      offset += rows;
+      const auto datum = check(arrow::compute::Take(batch, stride_index));
+      TENZIR_ASSERT(datum.kind() == arrow::Datum::Kind::RECORD_BATCH);
+      co_yield table_slice{
+        datum.record_batch(),
+        slice.schema(),
+      };
+    }
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.sample";
+  }
+
+  auto location() const -> operator_location override {
+    return operator_location::anywhere;
+  }
+
+  auto
+  optimize(expression const&, event_order) const -> optimize_result override {
+    // TODO: Consider adding an option that just subslices instead of sampling
+    // while respecting the input order. I.e., instead of taking every nth
+    // element we could also take all the elements from the front of every batch
+    // per batch.
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, sample_operator& x) -> bool {
+    return f.object(x)
+      .pretty_name("tenzir.plugins.sample.sample_operator")
+      .fields(f.field("args", x.args_));
+  }
+
+private:
+  operator_args args_{};
+};
+
+class plugin final : public virtual operator_plugin<sample_operator>,
+                     public virtual operator_factory_plugin {
+public:
+  auto signature() const -> operator_signature override {
+    return {.transformation = true};
+  }
+
+  auto operator_name() const -> std::string override {
+    return "sample";
+  }
+
+  auto
+  make(invocation inv, session ctx) const -> failure_or<operator_ptr> override {
+    auto str = std::optional<located<std::string>>{};
+    auto args = operator_args{};
+    TRY(argument_parser2::operator_("sample")
+          .add(args.period, "<duration>")
+          .add("mode", str)
+          .add("min_events", args.min_events)
+          .add("max_rate", args.max_rate)
+          .parse(inv, ctx));
+    if (args.period.inner.count() <= 0) {
```

Comment:
```suggestion
    if (args.period.inner < duration::zero()) {
```

Zero period = sample per batch. No reason not to allow this.

Reply:
We discussed this and agreed not to pursue this for now as it'd need special handling to avoid dividing by zero.